[DPE-8747] Handle persistent storage in set_up_database to prevent ObjectInUse errors

marceloneppel opened the pull request:
## Issue
The charm fails with "directory already in use as a tablespace" error when running on persistent storage for the `temp` storage after a reboot in the host machine.

## Solution
The issue occurred because the method assumed tmpfs storage and would rename the existing 'temp' tablespace and try to create a new one at the same location. With persistent storage, the directory remains in use by the renamed tablespace, causing PostgreSQL to reject the CREATE TABLESPACE command.

Changes:
- Add is_tmpfs() utility function to detect filesystem type via /proc/mounts
- Modify set_up_database() to differentiate between tmpfs and persistent storage:
  * tmpfs: Preserve existing behavior (rename + recreate for empty directory)
  * persistent: Only fix permissions, keep existing tablespace
- Extract tablespace handling into _handle_temp_tablespace_on_reboot() helper method to reduce cyclomatic complexity
- Add comprehensive test coverage (7 new tests for is_tmpfs, 1 for persistent storage handling)
- Update existing tests to mock filesystem detection This ensures the library works correctly with both tmpfs and persistent storage configurations for the temp tablespace.

Bumps version to 16.1.5.

## Checklist
- [ ] I have added or updated any relevant documentation.
- [ ] I have cleaned any remaining cloud resources from my accounts.

marceloneppel committed:
fix: handle persistent storage in set_up_database to prevent ObjectInUse errors

Fix a bug where set_up_database() fails with "directory already in use as a
tablespace" error when running on persistent storage after a reboot or failover.
The issue occurred because the method assumed tmpfs storage and would rename the
existing 'temp' tablespace and try to create a new one at the same location. With
persistent storage, the directory remains in use by the renamed tablespace, causing
PostgreSQL to reject the CREATE TABLESPACE command.
Changes:
- Add is_tmpfs() utility function to detect filesystem type via /proc/mounts
- Modify set_up_database() to differentiate between tmpfs and persistent storage:
  * tmpfs: Preserve existing behavior (rename + recreate for empty directory)
  * persistent: Only fix permissions, keep existing tablespace
- Extract tablespace handling into _handle_temp_tablespace_on_reboot() helper
  method to reduce cyclomatic complexity
- Add comprehensive test coverage (7 new tests for is_tmpfs, 1 for persistent
  storage handling)
- Update existing tests to mock filesystem detection
This ensures the library works correctly with both tmpfs and persistent storage
configurations for the temp tablespace.
Fixes cluster failures after reboot with persistent storage.
Bumps version to 16.1.5.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

onkolahmet commented on single_kernel_postgresql/utils/postgresql.py:
```diff
@@ -1107,39 +1107,85 @@ def _get_existing_databases(self) -> List[str]:
+            )
+        else:
+            # Persistent storage: Tablespace is still valid, permissions already fixed
+            # Log that we fixed permissions but didn't recreate
+            logger.info(
+                "Fixed permissions on temp tablespace directory (persistent storage), "
+                "existing tablespace remains valid"
+            )
+
```
Just a small thought, on persistent storage, we're assuming the existing tablespace is still valid. Should we add a sanity check that it's still at `temp_location`? Or is this guaranteed by design? In other words, could we include the actual `temp_location` path in the log message? Would make debugging easier.

marceloneppel replied:
Good suggestion. Added the `temp_location` to the log message on https://github.com/canonical/postgresql-single-kernel-library/pull/54/changes/32416d3cb7af52278a56efc32e68f85a1c9aa8ca.

Regarding "is this guaranteed by design": Yes, the tablespace location is guaranteed to be at temp_location by design. PostgreSQL maintains this in its catalog, and the charm always uses the same path. The edge case where the tablespace doesn't exist yet (reboot during initialization) is already handled by the early return at line 1121.

onkolahmet commented on single_kernel_postgresql/utils/postgresql.py:
```diff
@@ -1107,39 +1107,85 @@ def _get_existing_databases(self) -> List[str]:
+        if is_tmpfs(temp_location):
+            # tmpfs: Directory is empty after reboot, safe to rename and recreate
+            # Rename existing temp tablespace instead of dropping it.
+            new_name = f"temp_{datetime.now(timezone.utc).strftime('%Y%m%d%H%M%S')}"
+            cursor.execute(f"ALTER TABLESPACE temp RENAME TO {new_name};")
```
How do we handle if two renames somehow happen in the same second? Probably super rare, but maybe worth documenting?

marceloneppel replied:
Good suggestion. Added documentation explaining why timestamp collision cannot occur: https://github.com/canonical/postgresql-single-kernel-library/pull/54/commits/be8f00eade70c5d639656170eefa76ac21a07a5a.

Two renames cannot happen in the same second because consuming charms ensure leader-only execution (see [postgresql-operator example](https://github.com/canonical/postgresql-operator/blob/8fe608dc251e2d734cc0417aa415bfb119ec2d16/src/charm.py#L1327-1332)), and PostgreSQL serializes operations through exclusive locks on the tablespace object.

onkolahmet reviewed (commented):
LGTM! Just added some comments for clarity.

marceloneppel committed:
fix: include temp_location path in persistent storage log message

Add temp_location parameter to the log message in _handle_temp_tablespace_on_reboot()
when handling persistent storage scenarios. This provides better observability and
makes debugging easier by showing operators the exact directory path where permissions
were fixed.

Addresses feedback from PR review comment:
https://github.com/canonical/postgresql-single-kernel-library/pull/54#discussion_r2619109674

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
docs: explain timestamp collision prevention in tablespace rename

Add comment documenting why timestamp collision cannot occur during temp
tablespace rename operations on tmpfs storage.

The comment clarifies that consuming charms ensure leader-only execution and
PostgreSQL serializes operations through exclusive locks, making second-precision
timestamps sufficient for uniqueness.

Addresses feedback from PR review comment:
https://github.com/canonical/postgresql-single-kernel-library/pull/54#discussion_r2619118184

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
Merge remote-tracking branch 'origin/16/edge' into dpe-8747-fix-temp-tablespace-management

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>
